module/zfs/dsl_dataset.c

```diff
@@ -1565,6 +1561,40 @@ dsl_dataset_snapshot_check(void *arg, dmu_tx_t *tx)
 			rv = error;
 		}
 	}
+	if (rv != 0)
+		return (rv);
+
+	/* Now reserve space since we passed all the checks. */
+	for (pair = nvlist_next_nvpair(ddsa->ddsa_snaps, NULL);
```

Comment:
A second loop makes good sense, but it's unfortunate that we need to duplicate a lot of the work performed in the previous loop (without any error handling).

What if instead you restructured this to construct a list of all the held dataset's in the first loop (without releasing them).  Then you could remove each entry from the list and call `dsl_dataset_snapshot_reserve_space(); dsl_dataset_rele()`.  This would remove the need to call `dsl_dataset_hold()` in a second loop, and all the string handling rigmarole it requires.